test/test_rake_application.rb

```diff
@@ -634,4 +634,22 @@ def loader.make_dummy
     loader
   end
 
+  def test_running
+    was_running = false
+
+    @app.instance_eval do
+      app = self
+      intern(Rake::Task, "default").enhance do
+        was_running = app.running?
+      end
+    end
+
+    rakefile_default
+
+    assert !@app.running?
+    @app.run %w[--rakelib=""]
+    assert !@app.running?
+    assert was_running
+  end
+
```

Comment:
Could you also add a test case where the task execution fails but it correctly sets `@running` back to `false` after failure?